article/_posts/2020-07-02-What_Matters_in_RL.md

```diff
@@ -0,0 +1,137 @@
+
+## Training setup
+
+*Study*:
+
+> Number of parallel environments used (C1), the number of transitions gathered in each iteration, the number of passes over the data, minibatch size and how the data is split into minibatches
+
+*Results*:
+
+> Go over experience multiple times. Shuffle individual transitions before assigning them to minibatches and recompute advantages once per data pass. For faster wall-clock time training use many parallel environments and increase the batch size. Tune the number of transitions in each iteration if possible.
+
+## Timesteps handling 
+
+*Study*:
+
+> Discount factor, frame skip and how episode termination due to timestep limits are handled
+
+*Results*:
+
+> Discount factor $$\gamma$$ is one of the most important hyperparameters and should be tuned per environment (start with $$\gamma$$ = 0.99)
+
+## Optimizers
+
+*Study*:
+
+Adam vs RSMProp, their hyperparameters as well as learning rate decay
+
+*Results*:
+
+> Use Adam optimizer with momentum $$\beta 1$$ = 0.9 and a tuned learning rate (0.0003 is a safe default). Linearly decaying the learning rate may slightly improve performance but is of secondary importance.
+
+## Regularization
+
+*Study*:
+
+> We investigate different policy regularizers, which can have either the form of a penalty, (e.g. bonus for higher entropy) or a soft constraint (e.g. entropy should not be lower than some threshold) which is enforced with a Lagrange multiplier. In particular, we consider the following regularization terms: entropy, the Kullback–Leibler divergence (KL) between a reference $$N(0,1)$$ action distribution and the current policy and the KL divergence and reverse KL divergence between the current policy and the behavioral one as well as the “decoupled” KL divergence.
+
+*Results*:
+
+None.
+
+# Conclusions
+
+Many implementation details are available for a user wanting to implement their favorite RL algorithm. This paper provides some insight as to which should be chosen and how to achieve good performance.
+
+# Remarks
+- A reader of this review wanting to follow these recommendations is STRONGLY encouraged to go read the actual paper before making decisions based on this review.
+- The paper is only 8 pages long but comes with 40 pages of appendix detailing the implementation details and experiment design
+- See also: [The 32 Implementation Details of Proximal Policy Optimization (PPO) Algorithm](https://costa.sh/blog-the-32-implementation-details-of-ppo.html)
+- This is only for on-policy algorithms, but off-policy algorithms also suffer from similar problems. Recurrent RL is also very much obscure to this day.
+
+# References
+
+[^1]: See [https://vitalab.github.io/article/2020/01/14/Implementation_Matters.html](https://vitalab.github.io/article/2020/01/14/Implementation_Matters.html) for more info
+[^2]: Joelle Pineau: Reproducibility, Reusability, and Robustness in Deep Reinforcement Learning ICLR 2018, https://www.youtube.com/watch?v=Vh4H0gOwdIg
```

Comment:
Should be cross checked, but I think references need to have one blank line in between to be displayed on separate lines if that is the intention.